### ISSUE TITLE
Second derivative of Matern in zero is wrong

### ISSUE THREAD

**Issue by FelixBenning** [JuliaGaussianProcesses/KernelFunctions.jl#517]
```julia
julia> using KernelFunctions: MaternKernel

julia> k = MaternKernel(ν=5)
Matern Kernel (ν = 5, metric = Distances.Euclidean(0.0))

julia> import ForwardDiff as FD

julia> kx(x,y) = FD.derivative(t -> k(x+t, y), 0)
kx (generic function with 1 method)

julia> dk(x,y) = FD.derivative(t -> kx(x, y+t), 0)
dk (generic function with 1 method)

julia> dk(0,0)
0.0
```

This is because for a centered GP $Z$ with covariance function $k$

```math
dk(x,y) = \partial_x \partial_y k(x,y) = \partial_x \partial_y \mathbb{E}[Z(x),Z(y)] = \mathbb{E}[\partial_x Z(x) \partial_y Z(y)] = \text{Cov}(Z'(x), Z'(y))
```

And $\text{Cov}(Z'(0), Z'(0)) >0$.

$\nu=5$ should be plenty of space for numerical errors since this implies the GP is 5 times differentiable.

**Comment by FelixBenning** [comment]
This is most likely due to this

```julia
function _matern(ν::Real, d::Real)
    if iszero(d)
        return one(d)
    else
        y = sqrt(2ν) * d
        b = log(besselk(ν, y))
        return exp((one(d) - ν) * oftype(y, logtwo) - loggamma(ν) + ν * log(y) + b)
    end
end
```
https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/master/src/basekernels/matern.jl#L43-L45

The `if` block should not be constant but rather a taylor polynomial so that autodiff in this branch works. I am looking for a reference...

**Comment by devmotion** [comment]
Might be due to the hardcoded (constant) value for `x = y`. Hence possible the problem doesn't exist with ForwardDiff 0.11.

**Comment by devmotion** [comment]
> The if block should not be constant but rather a taylor polynomial so that autodiff in this branch works.

Shouldn't be necessary in ForwardDiff 0.11: It skips measure zero branches (which fixes some problems but broke existing code).

**Comment by FelixBenning** [comment]
https://github.com/cgeoga/BesselK.jl

https://arxiv.org/pdf/2201.00090.pdf

would be a possibility

**Comment by FelixBenning** [comment]
> Shouldn't be necessary in ForwardDiff 0.11: It skips measure zero branches (which fixes some problems but broke existing code).

wow neat - when is that going to be released?

**Comment by devmotion** [comment]
The package and paper is mainly concerned with derivatives wrt the order, which does not seem to be the issue in the OP.

> when is that going to be released?

I don't know. Initially the change was released in a non-breaking 0.10.X version but it broke a lot of downstream packages that rely on the current behaviour. So it was reverted and re-applied to the master branch but releases are only made in a separate 0.10.X branch without this change recently. I don't think anyone plans to release a 0.11 any time soon because the same thing will happen and nobody wants to invest the time to fix all the broken downstream code.

**Comment by Crown421** [comment]
I think the Matern kernel is in fact correct:
```
using KernelFunctions, Enzyme, Plots
k = Matern52Kernel()
mk(d) = k(1.0, 1.0 + d)

mk(0.1)

dr = range(0.0, 0.1; length=200)

p = plot(layout=(1, 2))
plot!(p[1], dr, mk.(dr), label="matern", legend=:topright)

dmk(x) = only(autodiff(Forward, mk, DuplicatedNoNeed, Duplicated(x, 1.0)))
only(dmk(0.1))

plot!(p[2], dr, dmk.(dr), label="matern", legend=:topright)
```
![image](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/assets/9107200/7d931a5c-4e43-46da-bfb4-1e3777be4ab9)

**Comment by FelixBenning** [comment]
> ~I think the Matern kernel is in fact correct:~ Made a mistake here, Matern is incorrect,but RationalQuadratic is fine

neat so switching to enzyme would at least fix the rational quadratic.

**Comment by Crown421** [comment]
I have been working on this for a bit, and it seems the issue is not the matern kernel, or that the Taylor Expansion is needed, but instead it seems to related to the Euclidean distance. In the following code I did it all by hand, and get
```julia
using KernelFunctions
using Enzyme
using Plots

k = Matern52Kernel()
mk(d) = k(1.0, 1.0 + d)

kappa(d) = (1 + sqrt(5) * d + 5 * d^2 / 3) * exp(-sqrt(5) * d)

r = range(0, 0.2, length=30)

begin
    dist1(x, y) = sqrt((x - y)^2)
    dist2(x, y) = abs(x - y)
    ck1(x, y) = kappa(dist1(x, y))
    ck2(x, y) = kappa(dist2(x, y))

    p = plot(layout=(2, 1))
    plot!(p[1], r, ck1.(1.0, 1.0 .+ r), label="dist1")
    plot!(p[1], r, ck2.(1.0, 1.0 .+ r), label="dist2")
    plot!(p[1], r, mk.(r), label="ref")

    dmkd1(x, y) = only(autodiff(
        Forward,
        yt -> only(autodiff_deferred(Forward, ck1, DuplicatedNoNeed, Duplicated(x, 1.0), yt)),
        DuplicatedNoNeed,
        Duplicated(y, 1.0)))
    dmkd2(x, y) = only(autodiff(
        Forward,
        yt -> only(autodiff_deferred(Forward, ck2, DuplicatedNoNeed, Duplicated(x, 1.0), yt)),
        DuplicatedNoNeed,
        Duplicated(y, 1.0)))

    plot!(p[2], r, dmkd1.(1.0, 1.0 .+ r), label="dist1")
    plot!(p[2], r, dmkd2.(1.0, 1.0 .+ r), label="dist2")
end
```
![image](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/assets/9107200/f64c3845-cb18-441a-92d8-b7b9b908a885)

Repeating the same with the distance works correctly, so there seems to be some weird issue with the chain rule. 
![image](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/assets/9107200/49421d52-9a92-492b-ab97-9e8ded3677c7)

**Comment by FelixBenning** [comment]
If I understan your code correctly, you reimplemented kappa (in the first plot you make sure that it results in the same function by comparing it to the reference implementation) and then you take the derivative of this kappa.

But your kappa does not have an `if iszero(d)` like the general matern implementation does cf.

https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/master/src/basekernels/matern.jl#L43-L45

As autodiff simply takes the derivative of the branch it finds itself in, it will take the derivative of this if case (if d is zero). And the derivative of a constant is zero.

**Comment by devmotion** [comment]
> As autodiff simply takes the derivative of the branch it finds itself in

That's not true in general. ForwardDiff#master is supposed to ignore branches of measure zero.

**Comment by FelixBenning** [comment]
> That's not true in general. ForwardDiff#master is supposed to ignore branches of measure zero.

But it doesn't at the moment and not in the forseeable future as I understood from the reaction when I asked around about it. And the point of my comment was to explain what the issue was at the moment (and this is probably it).

**Comment by Crown421** [comment]
> But your kappa does not have an `if iszero(d)` like the general matern implementation does 

From my understanding and experiments, the general Matern implementation doesn't matter for `Matern52`. It is a specialized implementation, and `_matern` never enters into it.

**Comment by Crown421** [comment]
Continuing to look into it, it seems the issues with `sqrt`. The Euclidean distance is (effectively) defined as `sqrt((x-y)^2)`, which then causes issues in `x==y`. See also [this long discussion](https://github.com/EnzymeAD/Enzyme/discussions/1351) in Enzyme.

**Comment by devmotion** [comment]
No, ForwardDiff uses its own definitions (functions with `Dual` arguments) and DiffRules. The `sqrt` example reminds me of https://github.com/JuliaDiff/DiffRules.jl/pull/100 (which reverted a ForwardDiff bug introduced by defining the derivative of `abs(x)` as `sign(x)` instead of `signbit(x) ? -one(x) : one(x)`. Clearly the only difference is how the AD derivative at 0 is defined (obviously the function is not differentiable at 0) - but setting it to 0 (which is a valid subderivative) breaks the hessian example in the linked PR (even on ForwardDiff#master). Hence generally I think it's better to define "derivatives" at non-differentiable points by evaluating them as the left and hand side derivatives than setting them to some constant (!) subderivative (as argued above in this issue here as well IIRC).

**Comment by Crown421** [comment]
@devmotion Do you think adding something like [this special rule](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/5127a26e24c7ef3b861f18614e114c230389c611/src/chainrules.jl#L4) would be the solution in this context?

**Comment by Crown421** [comment]
@wsmoses A bit of a side issue, but would it make sense to have that `API.printall!` function in the Enzyme documentation? I was initially looking for something like this and could not find it.

**Comment by wsmoses** [comment]
Oh yeah go for it, contributions welcome!
